crates/rpc/src/handlers/validator.rs

```diff
@@ -1,55 +1,77 @@
-use std::sync::Arc;
-
-use ream_bls::pubkey::pubkey_from_str;
+use ream_consensus::validator::Validator;
 use ream_storage::db::ReamDB;
+use serde::{Deserialize, Serialize};
 use serde_json::json;
 use warp::{
     http::status::StatusCode,
     reject::Rejection,
-    reply::{self, Reply, with_status},
+    reply::{Reply, with_status},
 };
 
 use super::{BeaconResponse, state::get_state_from_id};
+use crate::types::{
+    errors::ApiError,
+    id::{ID, ValidatorID},
+};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ValidatorData {
+    index: String,
+    balance: String,
+    status: String,
+    validator: Validator,
+}
+
+impl ValidatorData {
+    pub fn new(index: usize, balance: u64, status: String, validator: Validator) -> Self {
+        Self {
+            index: index.to_string(),
+            balance: balance.to_string(),
+            status,
+            validator,
+        }
+    }
+}
 
 pub async fn get_validator_from_state(
-    state_id: String,
-    validator_id: String,
-    db: Arc<ReamDB>,
+    state_id: ID,
+    validator_id: ValidatorID,
+    db: ReamDB,
 ) -> Result<impl Reply, Rejection> {
-    let validator_key = match pubkey_from_str(&validator_id) {
-        Ok(key) => key,
-        Err(_) => {
-            let error_body = reply::json(&json!({
-                "code": 400,
-                "message": format!("Invalid validator ID: {}", validator_id),
-            }));
-            return Ok(with_status(error_body, StatusCode::BAD_REQUEST));
+    let state = get_state_from_id(state_id, db).await?;
+
+    let (index, validator) = {
+        match validator_id.clone() {
```

Comment:
```suggestion
        match &validator_id {
```

Can't we avoid this clone?